src/python/pants/engine/target_test.py

```diff
@@ -0,0 +1,186 @@
+        assert sources_field.raw_value == raw_source_files
+        result: HaskellSourcesResult = run_rule(
+            hydrate_haskell_sources,
+            rule_args=[sources_field],
+            mock_gets=[
+                MockGet(
+                    product_type=Snapshot,
+                    subject_type=PathGlobs,
+                    mock=lambda _: Snapshot(
+                        directory_digest=EMPTY_DIRECTORY_DIGEST,
+                        files=hydrated_source_files,
+                        dirs=(),
+                    ),
+                )
+            ],
+        )
+        return result
+
+    # Normal field
+    expected_files = ("monad.hs", "abstract_art.hs", "abstract_algebra.hs")
+    assert (
+        hydrate_field(
+            raw_source_files=["monad.hs", "abstract_*.hs"], hydrated_source_files=expected_files
+        ).snapshot.files
+        == expected_files
+    )
+
+    # Test pre-hydration validation
+    with pytest.raises(ValueError) as exc:
+        hydrate_field(raw_source_files=[0, 1, 2], hydrated_source_files=())  # type: ignore[call-arg]
+    assert "Not all elements of the iterable have type" in str(exc)
+
+    # Test post-hydration validation
+    with pytest.raises(ValueError) as exc:
+        hydrate_field(raw_source_files=["*.js"], hydrated_source_files=("not_haskell.js",))
+    assert "Received non-Haskell sources" in str(exc)
+
+
+def test_has_fields() -> None:
+    class UnrelatedField(PrimitiveField):
+        alias: ClassVar = "unrelated"
+        raw_value: Optional[bool]
+
+        @memoized_property
+        def value(self) -> bool:
+            if self.raw_value is None:
+                return False
+            return self.raw_value
+
+    tgt = HaskellTarget({})
+    assert tgt.has_fields([]) is True
+    assert tgt.has_fields([HaskellGhcExtensions]) is True
+    assert tgt.has_fields([UnrelatedField]) is False
+    assert tgt.has_fields([HaskellGhcExtensions, UnrelatedField]) is False
+
+
+def test_field_hydration_is_lazy() -> None:
+    # No error upon creating the Target because validation does not happen until a call site
+    # hydrates the specific field.
+    tgt = HaskellTarget(
+        {HaskellGhcExtensions.alias: ["GhcExistentialQuantification", "DoesNotStartWithGhc"]}
+    )
+    # When hydrating, we expect a failure.
+    with pytest.raises(ValueError) as exc:
+        tgt.get(HaskellGhcExtensions).value
+    assert "must be prefixed by `Ghc`" in str(exc)
+
+
+def test_add_custom_fields() -> None:
+    class CustomField(PrimitiveField):
+        alias: ClassVar = "custom_field"
+        raw_value: Optional[bool]
+
+        @memoized_property
+        def value(self) -> bool:
+            if self.raw_value is None:
+                return False
+            return self.raw_value
+
+    union_membership = UnionMembership(OrderedDict({HaskellTarget: OrderedSet([CustomField])}))
```

Comment:
This is a confusing union, I think.

In general, union relationships (like Rust enums and Python unions) represent an "is-a" relationship (similar to subclassing.) `CustomField` "is-_not_-a" `HaskellTarget`.

I recognize that changing this will likely require a bit more boilerplate. But violating the is-a pattern is going to make unions more challenging to explain, I think.

Reply:
See the commit https://github.com/pantsbuild/pants/pull/9284/commits/8f3295530d0f4a209bf5744a6e0560b6a09eaa21. Originally, we more correctly modeled this as `is-a` by requiring that every class define a `PluginField` like `PythonLibraryField`. But, this both resulted in lots of boilerplate and was confusing, as Daniel first pointed out in https://github.com/pantsbuild/pants/pull/9284#discussion_r391805164.

I'm willing to go back to having `PluginField`, but don't think it's worth it. Beyond extra boilerplate, it forces target authors to consider the concept of additional fields being added, even though they are by definition never going to consume those extra fields in their rules. (They wouldn't know at the time of writing the rules that those fields exist, so they wouldn't know to consume them). Instead, now, all targets work automatically with plugin fields for no additional work.

We could do some type of generated field, like `HaskellTarget.PluginField`, but that seems even more confusing to me because it's not discoverable and it could trip up MyPy (not certain).

Reply:
Ok. But I still strongly believe that this union construction is problematic. If it is completely hidden from users, it might be less of an issue, but using unions uniformly will make them easier to teach, and make it easier to generate documentation for them later.

Reply:
If our focus is avoiding this violation of `is-a`, then I think I'd advocate the inner field `HaskellTarget.PluginField` (or something like that). 

I like that approach the most because it avoids the verbosity and forcing target authors to think about plugin fields, which, they never will end up actually consuming as explained in the above comment. 

It's easy for us to add that inner class in a followup.

Reply:
Yea, that would be much better. Thanks.